File: aplikasi_antrian/backend/internal/service/validation_service.go
package service

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"aplikasi_antrian/backend/internal/model"
	"aplikasi_antrian/backend/internal/repository"
)

var (
	nikRegex       = regexp.MustCompile(`^\d{16}$`)
	bpjsRegex      = regexp.MustCompile(`^\d{13}$`)
	queueCodeRegex = regexp.MustCompile(`^\d{12}$`)
	digitsRegex    = regexp.MustCompile(`\D`)
)

type ValidationService struct {
	repo *repository.ValidationRepository
}

func NewValidationService(repo *repository.ValidationRepository) *ValidationService {
	return &ValidationService{repo: repo}
}

func (s *ValidationService) ValidateNIK(input string) (model.NIKValidationResponse, error) {
	nik := digitsOnly(input)
	if !nikRegex.MatchString(nik) {
		return model.NIKValidationResponse{
			IsValid: false,
			Message: "NIK harus terdiri dari 16 digit angka.",
		}, nil
	}

	patient, err := s.repo.FindPatientByNIK(nik)
	if err != nil {
		return model.NIKValidationResponse{}, err
	}
	if patient == nil {
		return model.NIKValidationResponse{
			IsValid: false,
			Message: "NIK tidak ditemukan pada database.",
		}, nil
	}

	return model.NIKValidationResponse{
		IsValid:     true,
		Message:     "NIK valid.",
		PatientID:   patient.ID,
		QueueNumber: patient.QueueNumber,
		PatientName: patient.Name,
	}, nil
}

func (s *ValidationService) ValidateBPJSOrNIK(input string) (model.BPJSValidationResponse, error) {
	cleaned := digitsOnly(input)

	if nikRegex.MatchString(cleaned) {
		patient, err := s.repo.FindPatientByNIK(cleaned)
		if err != nil {
			return model.BPJSValidationResponse{}, err
		}
		if patient == nil {
			return model.BPJSValidationResponse{
				IsValid: false,
				Message: "NIK tidak ditemukan pada database.",
			}, nil
		}

		return model.BPJSValidationResponse{
			IsValid:     true,
			Message:     "NIK valid.",
			QueueNumber: patient.QueueNumber,
			PatientName: patient.Name,
		}, nil
	}

	if bpjsRegex.MatchString(cleaned) {
		patient, err := s.repo.FindPatientByBPJS(cleaned)
		if err != nil {
			return model.BPJSValidationResponse{}, err
		}
		if patient == nil {
			return model.BPJSValidationResponse{
				IsValid: false,
				Message: "Nomor BPJS tidak ditemukan pada database.",
			}, nil
		}

		return model.BPJSValidationResponse{
			IsValid:     true,
			Message:     "Nomor BPJS valid.",
			QueueNumber: patient.QueueNumber,
			PatientName: patient.Name,
		}, nil
	}

	return model.BPJSValidationResponse{
		IsValid: false,
		Message: "Input harus 16 digit NIK atau 13 digit nomor BPJS.",
	}, nil
}

func (s *ValidationService) ValidateQueueCode(input string) (model.QueueCodeValidationResponse, error) {
	queueCode := normalizeQueueCode(input)
	if !queueCodeRegex.MatchString(queueCode) {
		return model.QueueCodeValidationResponse{
			IsValid: false,
			Message: "Kode antrian harus 12 karakter (huruf/angka).",
		}, nil
	}

	record, err := s.repo.FindQueueByCode(queueCode)
	if err != nil {
		return model.QueueCodeValidationResponse{}, err
	}
	if record == nil {
		return model.QueueCodeValidationResponse{
			IsValid: false,
			Message: "Kode antrian tidak ditemukan pada database.",
		}, nil
	}

	return model.QueueCodeValidationResponse{
		IsValid: true,
		Message: "Kode antrian valid.",
		Data: &model.QueueVerificationData{
			QueueCode:    record.QueueCode,
			QueueNumber:  record.QueueNumber,
			PatientName:  record.PatientName,
			ClinicName:   record.ClinicName,
			DoctorName:   record.DoctorName,
			ScheduleInfo: record.ScheduleInfo,
			CreatedAt:    record.CreatedAt,
		},
	}, nil
}

func digitsOnly(input string) string {
	return digitsRegex.ReplaceAllString(input, "")
}

func normalizeQueueCode(input string) string {
	upper := strings.ToUpper(strings.TrimSpace(input))
	return regexp.MustCompile(`[^A-Z0-9]`).ReplaceAllString(upper, "")
}

func (s *ValidationService) CreatePharmacyQueue(patientID, patientName string) (model.PharmacyQueueResponse, error) {
	lastNumber, err := s.repo.GetLastPharmacyQueueNumber()
	if err != nil {
		return model.PharmacyQueueResponse{}, err
	}

	nextNumber := lastNumber + 1
	queueNumber := fmt.Sprintf("F%03d", nextNumber)

	// Generate pharmacy queue code
	pharmacyQueueCode := fmt.Sprintf("PHARM%03d", nextNumber)

	pharmacyQueue := &model.PharmacyQueue{
		ID:                fmt.Sprintf("PQ-%04d", nextNumber),
		PharmacyQueueCode: pharmacyQueueCode,
		QueueNumber:       queueNumber,
		PatientID:         patientID,
		ClinicName:        "FARMASI",
		DoctorName:        "-",
		ScheduleInfo:      "Pengambilan obat",
		CreatedAt:         time.Now(),
	}

	if err := s.repo.CreatePharmacyQueue(pharmacyQueue); err != nil {
		return model.PharmacyQueueResponse{}, err
	}

	return model.PharmacyQueueResponse{
		IsValid:     true,
		Message:     "Nomor antrian farmasi berhasil dibuat.",
		QueueNumber: queueNumber,
		PatientName: patientName,
		ClinicName:  "FARMASI",
	}, nil
}

File: aplikasi_antrian/lib/services/validation_service.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/validation_models.dart';

class ValidationService {
  ValidationService._();

  static const Duration _requestTimeout = Duration(seconds: 5);
  static const String _apiBaseUrl = String.fromEnvironment(
    'API_BASE_URL',
    defaultValue: 'http://10.0.2.2:8081',
  );

  static final RegExp _nikRegex = RegExp(r'^\d{16}$');
  static final RegExp _bpjsRegex = RegExp(r'^\d{13}$');
  static final RegExp _queueCodeRegex = RegExp(r'^[A-Z0-9]{12}$');

  static Future<NikValidationResponse> validateNik(String nikInput) async {
    final nik = _digitsOnly(nikInput);

    if (!_nikRegex.hasMatch(nik)) {
      return const NikValidationResponse(
        isValid: false,
        message: 'NIK harus terdiri dari 16 digit angka.',
      );
    }

    try {
      final json = await _postJson(
        '/api/v1/validate/nik',
        {'nik': nik},
      );
      return NikValidationResponse.fromJson(json);
    } on TimeoutException {
      return const NikValidationResponse(
        isValid: false,
        message: 'Validasi NIK timeout. Silakan coba lagi.',
      );
    } on _ValidationApiException catch (error) {
      return NikValidationResponse(
        isValid: false,
        message: error.message,
      );
    } catch (_) {
      return const NikValidationResponse(
        isValid: false,
        message: 'Tidak dapat terhubung ke API backend.',
      );
    }
  }

  static Future<BpjsValidationResponse> validateBpjsOrNik(String input) async {
    final cleaned = _digitsOnly(input);

    if (!_nikRegex.hasMatch(cleaned) && !_bpjsRegex.hasMatch(cleaned)) {
      return const BpjsValidationResponse(
        isValid: false,
        message: 'Input harus 16 digit NIK atau 13 digit nomor BPJS.',
      );
    }

    try {
      final json = await _postJson(
        '/api/v1/validate/bpjs-or-nik',
        {'input': cleaned},
      );
      return BpjsValidationResponse.fromJson(json);
    } on TimeoutException {
      return const BpjsValidationResponse(
        isValid: false,
        message: 'Validasi NIK/BPJS timeout. Silakan coba lagi.',
      );
    } on _ValidationApiException catch (error) {
      return BpjsValidationResponse(
        isValid: false,
        message: error.message,
      );
    } catch (_) {
      return const BpjsValidationResponse(
        isValid: false,
        message: 'Tidak dapat terhubung ke API backend.',
      );
    }
  }

  static Future<QueueCodeValidationResponse> validateQueueCode(
    String queueCodeInput,
  ) async {
    final queueCode = queueCodeInput
        .trim()
        .toUpperCase()
        .replaceAll(RegExp(r'[^A-Z0-9]'), '');

    if (!_queueCodeRegex.hasMatch(queueCode)) {
      return const QueueCodeValidationResponse(
        isValid: false,
        message: 'Kode antrian harus 12 karakter (huruf/angka).',
      );
    }

    try {
      final json = await _postJson(
        '/api/v1/validate/queue-code',
        {'queueCode': queueCode},
      );
      return QueueCodeValidationResponse.fromJson(json);
    } on TimeoutException {
      return const QueueCodeValidationResponse(
        isValid: false,
        message: 'Validasi kode antrian timeout. Silakan coba lagi.',
      );
    } on _ValidationApiException catch (error) {
      return QueueCodeValidationResponse(
        isValid: false,
        message: error.message,
      );
    } catch (_) {
      return const QueueCodeValidationResponse(
        isValid: false,
        message: 'Tidak dapat terhubung ke API backend.',
      );
    }
  }

  static Future<BpjsValidationResponse> generatePharmacyQueue({
    required String patientID,
    required String patientName,
  }) async {
    try {
      final json = await _postJson(
        '/api/v1/pharmacy/queue',
        {
          'patientId': patientID,
          'patientName': patientName,
        },
      );
      return BpjsValidationResponse.fromJson(json);
    } on TimeoutException {
      return const BpjsValidationResponse(
        isValid: false,
        message: 'Generate antrian farmasi timeout. Silakan coba lagi.',
      );
    } on _ValidationApiException catch (error) {
      return BpjsValidationResponse(
        isValid: false,
        message: error.message,
      );
    } catch (_) {
      return const BpjsValidationResponse(
        isValid: false,
        message: 'Tidak dapat terhubung ke API backend.',
      );
    }
  }

  static Future<Map<String, dynamic>> _postJson(
    String path,
    Map<String, dynamic> payload,
  ) async {
    final uri = Uri.parse('$_apiBaseUrl$path');
    final response = await http
        .post(
          uri,
          headers: const {
            'Content-Type': 'application/json',
          },
          body: jsonEncode(payload),
        )
        .timeout(_requestTimeout);

    Map<String, dynamic> jsonBody;
    try {
      final decoded = jsonDecode(response.body);
      if (decoded is! Map<String, dynamic>) {
        throw const _ValidationApiException('Response API tidak valid.');
      }
      jsonBody = decoded;
    } on FormatException {
      throw const _ValidationApiException('Response API tidak valid.');
    }

    if (response.statusCode >= 200 && response.statusCode < 300) {
      return jsonBody;
    }

    throw _ValidationApiException(
      jsonBody['message']?.toString() ??
          'Server mengembalikan error ${response.statusCode}.',
    );
  }

  static String _digitsOnly(String input) {
    return input.replaceAll(RegExp(r'\D'), '');
  }
}

class _ValidationApiException implements Exception {
  final String message;

  const _ValidationApiException(this.message);
}
